sdk/couchbase-core/tests/users.rs

```diff
@@ -0,0 +1,305 @@
+        create_and_ensure_user(&agent, &user).await;
+
+        delete_and_ensure_user(&agent, &username).await;
+
+        let opts = GetUserOptions::new(&username, "local");
+        let err = agent
+            .get_user(&opts)
+            .await
+            .expect_err("expected get after delete to error");
+
+        match err.kind() {
+            error::ErrorKind::Mgmt(e) => {
+                if let mgmtx::error::ErrorKind::Server(e, ..) = e.kind() {
+                    assert_eq!(e.kind(), &mgmtx::error::ServerErrorKind::UserNotFound);
+                } else {
+                    panic!("expected get after delete to error with UserNotFound");
+                }
+            }
+            _ => panic!("expected get after delete to error with UserNotFound"),
+        };
+    });
+}
+
+#[test]
+fn test_get_user() {
+    run_test(async |mut agent| {
+        let username = generate_key_with_letter_prefix();
+        let display_name = generate_key_with_letter_prefix();
+        let roles = vec![
+            Role::new("replication_target").bucket(&agent.test_setup_config.bucket),
+            Role::new("replication_admin"),
+        ];
+
+        let user = User::new(&username, display_name, roles).password("password");
+        create_and_ensure_user(&agent, &user).await;
+
+        let opts = GetUserOptions::new(&username, "local");
+        let actual = agent.get_user(&opts).await.unwrap();
+
+        assert_user(&user, &actual);
+    });
+}
+
+#[test]
+fn test_get_all_users() {
+    run_test(async |mut agent| {
+        let username = generate_key_with_letter_prefix();
+        let display_name = generate_key_with_letter_prefix();
+        let roles = vec![
+            Role::new("replication_target").bucket(&agent.test_setup_config.bucket),
+            Role::new("replication_admin"),
+        ];
+
+        let user = User::new(username, display_name, roles).password("password");
+        create_and_ensure_user(&agent, &user).await;
+
+        let opts = GetAllUsersOptions::new("local");
+        let users = agent.get_all_users(&opts).await.unwrap();
+
+        let mut actual = None;
+        for actual_user in users {
+            if actual_user.user.username == user.username {
+                actual = Some(actual_user);
+            }
+        }
+
+        assert_user(&user, actual.as_ref().unwrap());
+    });
+}
+
+fn assert_user(expected: &User, actual: &UserAndMetadata) {
+    assert_eq!(actual.domain, "local");
+    assert_eq!(2, actual.effective_roles.len());
+    assert!(actual.external_groups.is_empty());
+    assert_eq!(actual.user.username, expected.username);
+    assert_eq!(actual.user.display_name, expected.display_name);
+    assert_eq!(actual.user.groups, expected.groups);
+    assert_eq!(actual.user.roles, expected.roles);
+}
+
+async fn create_and_ensure_user(agent: &Agent, user: &User) {
+    agent
+        .upsert_user(&UpsertUserOptions::new(user, "local"))
+        .await
+        .unwrap();
+
+    loop {
```

Comment:
It seems like there's no way of leaving this loop when things go wrong? I assume there's a higher level timeout that would catch it, but I'm just wondering if that timeout would take a while to be hit, which could be a nuisance when these tests start failing?

Reply:
Yeah fair point. A general thing we have throughout our tests is a lack of timeouts at present. I'll update here and later do a thorough sweep to apply timeouts to operations.